public/app/features/serviceaccounts/ServiceAccountPage.tsx

```diff
@@ -55,6 +93,45 @@ const ServiceAccountPageUnconnected = ({
     deleteServiceAccountToken(parseInt(match.params.id, 10), key.id!);
   };
 
+  const onCreateToken = () => {
+    createServiceAccountToken(
+      serviceAccount.id,
+      {
+        name: newTokenName,
+        role: OrgRole.Viewer,
+      },
+      onTokenCreated
+    );
+  };
+
+  const onTokenCreated = (key: string) => {
+    setNewToken(key);
+  };
+
+  const closeModal = () => {
+    setIsModalOpen(false);
+    setNewTokenName('');
+    setIsWithExpirationDate(false);
+    setNewTokenExpirationDate('');
+    setIsExpirationDateValid(false);
+    setNewToken('');
```

Comment:
If there are a lot of state values, it might be worth considering `useReducer` hook instead of setting all of them individually via `setState`. One of the benefits is that all this call could've been done via dispatching one action.

Reply:
Yeah, it's not a requirement, but can be beneficial in some cases.

Reply:
I just moved modal component to separate file, so now it less messy, so, probably, it does not make sense to use reducers here.